akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehaviors4.scala

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (C) 2017-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.typed.scaladsl
+
+import scala.util.Try
+
+import akka.Done
+import akka.actor.typed.BackoffSupervisorStrategy
+import akka.actor.typed.Behavior.DeferredBehavior
+import akka.persistence._
+import akka.persistence.typed.EventAdapter
+
+/*
+API experiment with factory for command and event handler
+- commandHandler and eventHandler defined as functions as before, without enclosing class
+- to avoid repeating type parameters when several command handlers are defined a HandlerFactory
+  can be used
+- see AccountExample4
+*/
+
+object PersistentBehaviors4 {
+
+  class HandlerFactory[Command, Event, State] {
+    def commandHandler(handler: (State, Command) ⇒ Effect[Event, State]): (State, Command) ⇒ Effect[Event, State] =
+      handler
+
+    def subStateCommandHandler[S <: State](handler: (S, Command) ⇒ Effect[Event, State]): (S, Command) ⇒ Effect[Event, State] =
+      handler
+
+    /**
+     * Convenience for simple commands that don't need the state.
+     *
+     * @see [[Effect]] for possible effects of a command.
+     */
+    def command(handler: Command ⇒ Effect[Event, State]): (State, Command) ⇒ Effect[Event, State] =
+      (_, cmd) ⇒ handler(cmd)
+
+    def eventHandler(handler: (State, Event) ⇒ State): (State, Event) ⇒ State =
+      handler
+  }
+
+  /**
+   * Type alias for the command handler function that defines how to act on commands.
+   *
+   * The type alias is not used in API signatures because it's easier to see (in IDE) what is needed
+   * when full function type is used. When defining the handler as a separate function value it can
+   * be useful to use the alias for shorter type signature.
+   */
+  type CommandHandler[Command, Event, State] = (State, Command) ⇒ Effect[Event, State]
+
+  /**
+   * Type alias for the command handler function for a subclass of the `State` that defines how to act on commands.
+   *
+   * The type alias is not used in API signatures because it's easier to see (in IDE) what is needed
+   * when full function type is used. When defining the handler as a separate function value it can
+   * be useful to use the alias for shorter type signature.
+   */
+  type SubStateCommandHandler[Command, Event, State, S <: State] = (S, Command) ⇒ Effect[Event, State]
```

Comment:
I guess we don't need this here also if we change the above one to use `S <: State`